src/codegen.cpp

```diff
@@ -4480,6 +4490,98 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
         // it to the inferred type.
         return mark_julia_type(ctx, val, true, (jl_value_t*)jl_any_type);
     }
+    else if (head == new_opaque_closure_sym) {
+        size_t nargs = jl_array_len(ex->args);
+        if (nargs < 4) {
+            emit_error(ctx, "Not enough arguments in new_opaque_closure");
+            return jl_cgval_t();
+        }
+        jl_cgval_t *argv = (jl_cgval_t*)alloca(sizeof(jl_cgval_t) * nargs);
+        for (size_t i = 0; i < nargs; ++i) {
+            argv[i] = emit_expr(ctx, args[i]);
+        }
+        const jl_cgval_t &argt = argv[0];
+        const jl_cgval_t &ub = argv[2];
+        const jl_cgval_t &source = argv[3];
+        if (source.constant == NULL || argt.constant == NULL || ub.constant == NULL) {
+            // For now, we require non-constant source to be created using
+            // eval - we may lift this restriction if necessary.
+            emit_error(ctx, "Opaque closure source be constant");
+            return jl_cgval_t();
+        }
+        if (jl_is_code_info(source.constant)) {
+            // TODO: Emit this inline and outline it late using LLVM's coroutine
+            // support.
```

Comment:
coroutine support would allow llvm passes to optimize the capture list. We could write our own version of it, but I think the existing one would work fine.